### PR TITLE
add --fix support to `cargo-clippy`

### PR DESCRIPTION
**Pull request by yaahc**
Prior to this we had started work on integrating clippy as a subcommand directly into cargo in the form of `cargo clippy-preview` and `cargo fix --clippy`. In the course of that work it was decided that the best approach would be to strictly add the features clippy needed to cargo in order to insert `clippy-driver` only for workspace crates. This was accomplished by adding a `RUSTC_WORKSPACE_WRAPPER` env variable to cargo that will override the normal `RUSTC_WRAPPER` when both are present and the current crate is a workspace crate.

This change adds support to clippy to use this by setting the `RUSTC_WORKSPACE_WRAPPER` env variable instead `RUSTC_WRAPPER` and by detecting `--fix` as an arg and swapping out the `check` cargo command for `fix` when it is present.

WIP, here are the current issues that I still need to resolve

- [x] Detect if we're running on nightly rust
  - [x] Set `RUSTC_WORKSPACE_WRAPPER` on nightly, and `RUSTC_WRAPPER` on stable
  - [x] Error out on stable when `--fix` is specified, because stable currently hasn't landed the PR for `RUSTC_WORKSPACE_WRAPPER` so if we set this it just runs check and silently fails
- [ ] Update the help text
  - [ ] The current plan is to shell out to `cargo check --help` and then postprocess the output to mention clippy instead of check where appropriate and to add the extra info about `--fix` and the `-- -A lint` options.
- [x] tests?

changelog: add `--fix` arg to `cargo-clippy`

### CODE REVIEW

**Comment by yaahc** [comment]
I'm worried about the `nightly` vs `stable` checking, and how to make sure that maintains correctness as `cargo-clippy` lands on the next beta/stable versions, once the current version of `cargo` is stable we might be able to remove the special case error message for `--fix` and just let `cargo` print the error message for us when we set the env variable on stable / without -Zunstable-options

Also I don't think we ever need to check for `-Zunstable-options` because cargo will handle printing that diagnostic nicely for us when its missing / needed.

**Comment by yaahc** [comment]
Assuming that we will only ever invoke the nightly version of `cargo-clippy` when on nightly then this is probably all moot and we can just ignore the version checking except for the need to stick to `RUSTC_WRAPPER` until cargo stabilizes `RUSTC_WORKSPACE_WRAPPER`

**Comment by flip1995** [comment]
About the nightly check: We could ship it without this check, since it will land on stable together with the current cargo. That means, that rustup users shouldn't trip over this change. If Clippy was compiled and installed from source, this might be a problem depending on the cargo version, that is used.

**Comment by yaahc** [comment]
I think if we merged it without the nightly check it would break clippy on stable if cargo doesn't stabilize `RUSTC_WORKSPACE_WRAPPER` before the next stable cutoff because we would try to do `RUSTC_WORKSPACE_WRAPPER=clippy-driver cargo check ...` which would still be invalid on anything other than nightly + unstable-options

**Comment by flip1995** [comment]
Ah `RUSTC_WORKSPACE_WRAPPER` isn't stabilized in cargo. Sorry I missed that. Yeah in that case, we'll need that check.

**Comment by flip1995** [comment]
I don't know, if or how we can check if nightly cargo is used, so I guess we'll have to wait, until `RUSTC_WORKSPACE_WRAPPER` gets stabilized (?).

**Comment by yaahc** [comment]
I'll ping @ehuss to see if he has any ideas.

**Comment by matthiaskrgr** [comment]
>I don't know, if or how we can check if nightly cargo is used

Perhaps the [rustc_version](https://crates.io/crates/rustc_version) crate could be helpful here?

**Comment by ehuss** [comment]
The intent was to have this as an unstable feature in clippy for a short while, and then stabilize it on both sides at some point.  So you would specify `cargo +nightly clippy -Zunstable-options --fix`. 

I was initially thinking there might be some way for `cargo-clippy` to detect if it is on nightly, but thinking about it more, it may not be necessary?  Perhaps it could work by just checking if `-Zunstable-options` is in args, and if it is, enable the use of `RUSTC_WORKSPACE_WRAPPER`.  It would pass `-Zunstable-options` to cargo which would handle checking if it is actually on nightly.

How does that sound?

**Comment by yaahc** [comment]
Sounds like a perfect solution

**Comment by yaahc** [comment]
Okay so this works now, wondering how much we care about getting the tests / help text change finished before merging this, given that its behind unstable options I'm thinking I should try to add one test to ensure that when we enable unstable options we call cargo with `RUSTC_WORKSPACE_WRAPPER` and then leave the testing fix integration and updating the help text to a follow up PR.

**Comment by flip1995** [comment]
I agree. A test for this would be good. The help can be adapted later.

**Comment by shepmaster** [comment]
I merged this branch with master and figured out how to compile my own Clippy, all because this type of thing is **really** valuable for revitalizing old projects (some of mine were started before Rust 1.0!). Looking forward to this being generally available.

**Comment by flip1995** [comment]
We could merge this and revisit testing this later. I'll take a final look at this later and then add it to the `S-waiting-on-rustup` list.

File: src/main.rs [hunk 1]
```diff
@@ -58,13 +58,38 @@ where
...
         args.push(arg);
     }
 
+    if fix {
+        if unstable_options {
+            args[0] = "fix".to_owned();
+        } else {
+            panic!("Usage of `--fix` requires `-Z unstable-options`");
+        }
+    }
```
Comment (phansch) [inline src/main.rs:85]: The argument building from the start of the function until here could be extracted into a separate function and unit tested maybe?

Comment (yaahc) [reply src/main.rs:85]: will do

**Comment by yaahc** [comment]
Also I know that the formatting is wrong but nightly rustfmt isnt available on the current nightly which is the only one that builds clippy soooooooooooooo /shrug

**Comment by flip1995** [comment]
@phansch can you re-review this please, I still have 5 PRs in my review queue.

File: src/main.rs [hunk 2]
```diff
@@ -37,68 +41,135 @@ fn show_version() {
...
+}
+
+impl ClippyCmd
 {
-    let mut args = vec!["check".to_owned()];
+    fn new<I>(mut old_args: I) -> Self
+    where
+        I: Iterator<Item = String>,
+    {
+        let mut cmd = "check";
```
Comment (phansch) [inline src/main.rs:72]: Maybe `cargo_subcommand` is more descriptive here?
```suggestion
        let mut cargo_subcommand = "check";
```

File: src/main.rs [hunk 3]
```diff
@@ -37,68 +41,130 @@ fn show_version() {
...
+        if self.unstable_options {
+            "RUSTC_WORKSPACE_WRAPPER"
+        } else {
+            "RUSTC_WRAPPER"
         }
-        args.push(arg);
     }
 
-    let clippy_args: String = old_args.map(|arg| format!("{}__CLIPPY_HACKERY__", arg)).collect();
+    fn path(&self) -> PathBuf {
```
Comment (flip1995) [inline src/main.rs:118]:
```suggestion
    fn path() -> PathBuf {
```

File: src/main.rs [hunk 4]
```diff
@@ -37,68 +41,130 @@ fn show_version() {
...
+            })
+            .map(|p| ("CARGO_TARGET_DIR", p))
     }
 
-    let exit_status = std::process::Command::new("cargo")
-        .args(&args)
-        .env("RUSTC_WRAPPER", path)
-        .env("CLIPPY_ARGS", clippy_args)
-        .envs(target_dir)
+    fn to_std_cmd(self) -> Command {
```
Comment (flip1995) [inline src/main.rs:146]:
```suggestion
    fn into_std_cmd(self) -> Command {
```

File: src/main.rs [hunk 5]
```diff
@@ -110,3 +176,44 @@ where
         Err(exit_status.code().unwrap_or(-1))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```
Comment (flip1995) [inline src/main.rs:182]:
```suggestion
    use super::ClippyCmd;
```

File: src/main.rs [hunk 6]
```diff
@@ -110,3 +176,44 @@ where
...
+
+    #[test]
+    fn fix_unstable() {
+        let args = "cargo clippy --fix -Zunstable-options"
+            .split_whitespace()
+            .map(ToString::to_string);
+        let cmd = ClippyCmd::new(args);
+        assert_eq!("fix", cmd.cargo_subcommand);
+        assert_eq!("RUSTC_WORKSPACE_WRAPPER", cmd.path_env());
+        assert!(cmd.args.iter().find(|arg| arg.ends_with("unstable-options")).is_some());
```
Comment (flip1995) [inline src/main.rs:199]:
```suggestion
        assert!(cmd.args.iter().any(|arg| arg.ends_with("unstable-options")));
```

File: src/main.rs [hunk 7]
```diff
@@ -37,68 +41,130 @@ fn show_version() {
...
 
-    let exit_status = std::process::Command::new("cargo")
-        .args(&args)
-        .env("RUSTC_WRAPPER", path)
-        .env("CLIPPY_ARGS", clippy_args)
-        .envs(target_dir)
+    fn to_std_cmd(self) -> Command {
+        let mut cmd = Command::new("cargo");
+
+        cmd.env(self.path_env(), self.path())
```
Comment (flip1995) [inline src/main.rs:149]:
```suggestion
        cmd.env(self.path_env(), Self::path())
```

File: src/main.rs [hunk 8]
```diff
@@ -37,68 +41,130 @@ fn show_version() {
...
+    }
+}
+
+fn process<I>(old_args: I) -> Result<(), i32>
+where
+    I: Iterator<Item = String>,
+{
+    let cmd = ClippyCmd::new(old_args);
+
+    let mut cmd = cmd.to_std_cmd();
```
Comment (flip1995) [inline src/main.rs:165]:
```suggestion
    let mut cmd = cmd.into_std_cmd();
```

**Comment by flip1995** [comment]
@bors r=phansch,flip1995

Thanks!

**Pull request merged**